### ISSUE TITLE
After running tests R.Host.exe processes do not terminate

### ISSUE THREAD

**Issue by MikhailArkhipov** [Microsoft/RTVS#3209]
There is no client so they should be auto-terminating I think... Should they?

`is_terminating` is true, `response_state == RESPONSE_EXPECTED`

but

```
            for (;;) {
                message msg;
                for (;;) { 
```

never terminates.

**Comment by Neil-Schneider** [comment]
I am running 1.0-rc1 and I have a similar issue, although it is not when running tests. When "resetting" the R session the R.Host.exe processes does not terminate. I first notice this because the old session was using ~1.5CPUs in the background. 

1. I can replicate the failure to terminate by clicking the "reset" button.
2. I have has the old session processing in the background twice, but have not figured out how to replicate it.

Let me know if I can provide you with more information, or if you would like me to open this in a separate issue.

**Comment by MikhailArkhipov** [comment]
RTVS can actually have up to 3 sessions running. This is normal. 

1. Your actual interactive session
2. Intellisense information session. It is separate because otherwise when (1) runs long calculation you'd lose completion in the editor.
3. Package manager session.

Test case is different since in unit tests host process may get created differently or test may fail and terminate early causing stale sessions.

**Comment by Neil-Schneider** [comment]
Thanks. I just noticed that a second R Host was running without the "reset".

**Comment by MikhailArkhipov** [comment]
This is generally normal. Reset quits all sessions and restarts primary interactive one. Later, when intellisense needs more data (say, signature information for a function that is not yet in the cache), it will start its own session.

**Comment by int19h** [comment]
Reset doesn't actually terminate all sessions, it only does that to the REPL session.

**Comment by int19h** [comment]
As far as reproing it with tests, what I'm seeing is that we leave a lot of _broker_ processes around. This can in turn produce stray host processes (although I haven't observed it yet). Host will auto-exit if its stdin/stdout pipe is closed, which happens automatically if the broker (which holds the other end of the pipe) terminates for any reason. On the other hand, broker is watching the process that launched it via PID, and auto-exits if that parent process exits.

In case of tests, however, the parent process is vstest.executionengine.*.exe, which is *not* terminated when the test run completes (it's kept around by VS to be reused if you run more tests). So all that auto-exit functionality is not actually engaged.

So far as I can see, the real problem is that some of the tests don't properly dispose `RSessionProvider` (which disposes `LocalBrokerClient`, which ensures that the broker process is terminated, which would then also ensure that all host processes are terminated, even if the hosts themselves haven't been terminated explicitly for whatever reason).

**Comment by AlexanderSher** [comment]
If `RSessionProvider` is created by `RInteractiveWorkflow`, it should be disposed with it.
Have you found any specific tests?

**Comment by MikhailArkhipov** [comment]
See also my comment above. when I attached to the orphaned host, I saw that its termination flag was set to true but the loop never exited.

**Comment by int19h** [comment]
Some tests do not use `RInteractiveWorkflow`, and work with sessions and session providers directly.

All tests with `Intellisense` in their names seem to be having this effect.

One other thing of interest is that these tests do call `RSessionProvider.RemoveBrokerAsync` (coming from `PackageIndexUtility.DisposeAsync`). However, the implementation of this method does not call `Dispose` on the old broker.

**Comment by AlexanderSher** [comment]
> However, the implementation of this method does not call `Dispose` on the old broker.

That is definitely a bug. Though it doesn't explain what @MikhailArkhipov has found.

**Comment by int19h** [comment]
Do you mean `is_terminating` static variable inside `terminate_if_disconnected`? If you were able to view its value, the function must have been on the stack, right? I would very much like to see a call stack of the repro (but so far I couldn't get it into a situation where a host would linger).

**Comment by MikhailArkhipov** [comment]
the static one

**Comment by int19h** [comment]
If you can get a repro again, please send me the call stack and save a full heap dump. I'll keep trying to repro locally, but no luck so far.

**Comment by MikhailArkhipov** [comment]
I ran the entire stack of tests, some fail b/c #3009. I think failed tests left orphans. But it is still unclear how the loop above is supposed to exit...

**Comment by int19h** [comment]
`R_WaitEvent` has a Win32 message loop inside. `unblock_message_loop` is a helper function that posts dummy `WM_NULL` messages to unblock that loop, and is registered to be invoked when the transport is disconnected. 

One thing that I've noticed is that there's a potential race condition related to setting a flag `is_waiting_for_wm` to indicate that `WM_NULL` needs to be posted. It is currently possible for a disconnect to happen while the flag is not set (so message won't be posted), but when calling `R_WaitEvent` is imminent. This would cause the observed behavior. Alex has intermittent repros of this as well, and we're currently checking to see if fixing this issue makes them go away.

**Comment by MikhailArkhipov** [comment]
I got several running now. I didn't run tests, these are all from

Run F5
Debug | Terminate All.

```
 	win32u.dll!00007ffbf3fb1224()	Unknown
>	Microsoft.R.Host.exe!rhost::host::send_request_and_get_response::__l7::<lambda>(void * __formal) Line 833	C++
 	Microsoft.R.Host.exe!void <lambda>(void *)::<lambda_invoker_cdecl>(void * __p1) Line 835	C++
 	R.dll!000000006c781228()	Unknown
 	Microsoft.R.Host.exe!rhost::host::send_request_and_get_response(const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & name, const std::vector<picojson::value,std::allocator<picojson::value> > & args) Line 838	C++
 	Microsoft.R.Host.exe!rhost::r_util::send_request_and_get_response::__l2::<lambda>() Line 297	C++
 	Microsoft.R.Host.exe!rhost::host::with_cancellation<SEXPREC * <lambda>(void) >(rhost::r_util::send_request_and_get_response::__l2::SEXPREC * <lambda>(void) body) Line 55	C++
 	Microsoft.R.Host.exe!send_request_and_get_response(SEXPREC * name_sexp, SEXPREC * args_sexp) Line 322	C++
 	R.dll!000000006c794a20()	Unknown
 	R.dll!000000006c7e6ece()	Unknown
 	R.dll!000000006c7e95b7()	Unknown
 	R.dll!000000006c7e6ce1()	Unknown
 	R.dll!000000006c7e85f7()	Unknown
 	R.dll!000000006c7e6b13()	Unknown
 	R.dll!000000006c7e95b7()	Unknown
 	R.dll!000000006c7e6ce1()	Unknown
 	R.dll!000000006c7e85f7()	Unknown
 	R.dll!000000006c7e6b13()	Unknown
 	R.dll!000000006c88f648()	Unknown
 	R.dll!000000006c7e6ce1()	Unknown
 	R.dll!000000006c7e95b7()	Unknown
 	R.dll!000000006c7e6ce1()	Unknown
 	R.dll!000000006c7e85f7()	Unknown
 	R.dll!000000006c7e6b13()	Unknown
 	R.dll!000000006c7e719e()	Unknown
 	R.dll!000000006c7e6f29()	Unknown
 	R.dll!000000006c7ea791()	Unknown
 	R.dll!000000006c816421()	Unknown
 	R.dll!000000006c7dddb4()	Unknown
 	R.dll!000000006c7e69a1()	Unknown
 	R.dll!000000006c7e85f7()	Unknown
 	R.dll!000000006c7e6b13()	Unknown
 	R.dll!000000006c7e719e()	Unknown
 	R.dll!000000006c7e6a3c()	Unknown
 	R.dll!000000006c7dc400()	Unknown
 	R.dll!000000006c7e69a1()	Unknown
 	R.dll!000000006c7e85f7()	Unknown
 	R.dll!000000006c7e6b13()	Unknown
 	R.dll!000000006c7e95b7()	Unknown
 	R.dll!000000006c7e6ce1()	Unknown
 	R.dll!000000006c7e85f7()	Unknown
 	R.dll!000000006c7e6b13()	Unknown
 	Microsoft.R.Host.exe!rhost::r_util::init::__l2::void <lambda>(void)::__l5::<lambda>() Line 598	C++
 	Microsoft.R.Host.exe!rhost::util::r_top_level_exec::__l3::<lambda>(void * arg) Line 170	C++
 	Microsoft.R.Host.exe!void <lambda>(void *)::<lambda_invoker_cdecl>(void * arg) Line 170	C++
 	R.dll!000000006c781228()	Unknown
 	Microsoft.R.Host.exe!rhost::util::r_top_level_exec<void <lambda>(void) >(rhost::r_util::init::__l2::void <lambda>(void)::__l5::void <lambda>(void) protected_eval, const char * log_error_prefix) Line 170	C++
 	Microsoft.R.Host.exe!rhost::r_util::init::__l2::<lambda>() Line 600	C++
 	Microsoft.R.Host.exe!boost::detail::function::void_function_obj_invoker0<void <lambda>(void),void>::invoke(boost::detail::function::function_buffer & function_obj_ptr) Line 160	C++
 	Microsoft.R.Host.exe!boost::function0<void>::operator()() Line 772	C++
 	Microsoft.R.Host.exe!boost::signals2::detail::call_with_tuple_args<boost::signals2::detail::void_type>::m_invoke<boost::function<void __cdecl(void)> >(boost::function<void __cdecl(void)> & func, boost::signals2::detail::unsigned_meta_array<> __formal, const std::tuple<> & __formal, void * __formal) Line 118	C++
 	Microsoft.R.Host.exe!boost::signals2::detail::call_with_tuple_args<boost::signals2::detail::void_type>::operator()<boost::function<void __cdecl(void)>,0>(boost::function<void __cdecl(void)> & func, const std::tuple<> & args, boost::mpl::size_t<0> __formal) Line 90	C++
 	Microsoft.R.Host.exe!boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>::operator()<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > >(const boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > & connectionBody) Line 133	C++
 	Microsoft.R.Host.exe!boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> >::dereference() Line 110	C++
 	Microsoft.R.Host.exe!boost::iterators::iterator_core_access::dereference<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > >(const boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > & f) Line 550	C++
 	Microsoft.R.Host.exe!boost::iterators::detail::iterator_facade_base<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> >,boost::signals2::detail::void_type,boost::iterators::single_pass_traversal_tag,boost::signals2::detail::void_type & __ptr64,__int64,0,0>::operator*() Line 656	C++
 	Microsoft.R.Host.exe!boost::signals2::optional_last_value<void>::operator()<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > >(boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > first, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > last) Line 57	C++
 	Microsoft.R.Host.exe!boost::signals2::detail::combiner_invoker<void>::operator()<boost::signals2::optional_last_value<void>,boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > >(boost::signals2::optional_last_value<void> & combiner, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > first, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type>,std::_List_iterator<std::_List_val<std::_List_simple_types<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > > > >,boost::signals2::detail::connection_body<std::pair<enum boost::signals2::detail::slot_meta_group,boost::optional<int> >,boost::signals2::slot<void __cdecl(void),boost::function<void __cdecl(void)> >,boost::signals2::mutex> > last) Line 64	C++
 	Microsoft.R.Host.exe!boost::signals2::detail::signal_impl<void __cdecl(void),boost::signals2::optional_last_value<void>,int,std::less<int>,boost::function<void __cdecl(void)>,boost::function<void __cdecl(boost::signals2::connection const & __ptr64)>,boost::signals2::mutex>::operator()() Line 242	C++
 	Microsoft.R.Host.exe!boost::signals2::signal<void __cdecl(void),boost::signals2::optional_last_value<void>,int,std::less<int>,boost::function<void __cdecl(void)>,boost::function<void __cdecl(boost::signals2::connection const & __ptr64)>,boost::signals2::mutex>::operator()() Line 724	C++
 	Microsoft.R.Host.exe!rhost::host::terminate_if_disconnected() Line 205	C++
 	Microsoft.R.Host.exe!rhost::host::shutdown_if_requested() Line 212	C++
 	Microsoft.R.Host.exe!rhost::host::do_r_callback(bool allow_eval_interrupt) Line 874	C++
 	Microsoft.R.Host.exe!CallBack() Line 906	C++
 	R.dll!000000006c71fd6e()	Unknown
 	Microsoft.R.Host.exe!rhost::host::send_request_and_get_response::__l7::<lambda>(void * __formal) Line 835	C++
 	Microsoft.R.Host.exe!void <lambda>(void *)::<lambda_invoker_cdecl>(void * __p1) Line 835	C++
 	R.dll!000000006c781228()	Unknown
 	Microsoft.R.Host.exe!rhost::host::send_request_and_get_response(const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & name, const std::vector<picojson::value,std::allocator<picojson::value> > & args) Line 838	C++
 	Microsoft.R.Host.exe!rhost::host::send_request_and_get_response<std::vector<picojson::value,std::allocator<picojson::value> >,double,bool,picojson::value,picojson::value>(const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & name, const std::vector<picojson::value,std::allocator<picojson::value> > & <args_0>, const double & <args_1>, const bool & <args_2>, const picojson::value & <args_3>, const picojson::value & <args_4>) Line 83	C++
 	Microsoft.R.Host.exe!rhost::host::R_ReadConsole::__l2::<lambda>() Line 971	C++
 	Microsoft.R.Host.exe!rhost::host::with_cancellation<int <lambda>(void) >(rhost::host::R_ReadConsole::__l2::int <lambda>(void) body) Line 55	C++
 	Microsoft.R.Host.exe!R_ReadConsole(const char * prompt, char * buf, int len, int addToHistory) Line 1000	C++
 	R.dll!000000006c806ae2()	Unknown
 	R.dll!000000006c806d42()	Unknown
 	R.dll!000000006c806dd2()	Unknown
 	Microsoft.R.Host.exe!rhost::run(rhost::command_line_args & args) Line 259	C++
 	Microsoft.R.Host.exe!rhost::run(int argc, char * * argv) Line 264	C++
 	Microsoft.R.Host.exe!main(int argc, char * * argv) Line 272	C++
 	Microsoft.R.Host.exe!invoke_main() Line 65	C++
 	Microsoft.R.Host.exe!__scrt_common_main_seh() Line 253	C++
 	Microsoft.R.Host.exe!__scrt_common_main() Line 296	C++
 	Microsoft.R.Host.exe!mainCRTStartup() Line 17	C++
 	kernel32.dll!00007ffbf7008364()	Unknown
 	ntdll.dll!00007ffbf71170d1()	Unknown

```
